File: src/domain/money/format.test.ts
import { describe, expect, it } from 'vitest';
import { formatCompact, formatMoney, parseMoney } from './format';

describe('formatMoney', () => {
  it('usa formato colombiano con punto de miles', () => {
    expect(formatMoney(2_500_000)).toBe('$ 2.500.000');
  });

  it('no muestra decimales', () => {
    expect(formatMoney(9_900)).toBe('$ 9.900');
    expect(formatMoney(175_094)).toBe('$ 175.094');
  });

  it('maneja cero y negativos', () => {
    expect(formatMoney(0)).toBe('$ 0');
    expect(formatMoney(-48_000)).toBe('-$ 48.000');
  });
});

describe('parseMoney', () => {
  it('acepta lo que el usuario realmente escribe', () => {
    expect(parseMoney('85000')).toBe(85_000);
    expect(parseMoney('85.000')).toBe(85_000);
    expect(parseMoney('$ 85.000')).toBe(85_000);
    expect(parseMoney('1.500.000')).toBe(1_500_000);
  });

  it('devuelve null cuando no hay numero', () => {
    expect(parseMoney('')).toBeNull();
    expect(parseMoney('abc')).toBeNull();
  });
});

describe('formatCompact', () => {
  it('abrevia millones y miles', () => {
    expect(formatCompact(2_500_000)).toBe('$ 2,5 M');
    expect(formatCompact(85_000)).toBe('$ 85 k');
  });
});

describe('setMoneyLocale', () => {
  it('cambia la moneda de todos los formatos sin tocar los call sites', async () => {
    const { setMoneyLocale, currencySymbol } = await import('./format');
    try {
      setMoneyLocale('en-US', 'USD');
      expect(formatMoney(2500)).toBe('$2,500');
      expect(currencySymbol()).toBe('$');

      setMoneyLocale('es-ES', 'EUR');
      expect(formatMoney(2500)).toContain('€');
      expect(currencySymbol()).toBe('€');
      expect(formatCompact(2_500_000)).toBe('€ 2,5 M');
    } finally {
      // Otros tests asumen COP: dejar el modulo como estaba.
      setMoneyLocale('es-CO', 'COP');
    }
  });

  it('vuelve a colombiano al restaurar', () => {
    expect(formatMoney(2_500_000)).toBe('$ 2.500.000');
  });
});

/**
 * La misma tabla vive en mobile/lib/domain/money/format.dart y su test
 * afirma exactamente estos valores. Si cambia uno, que falle el otro.
 */
describe('paridad con la app nativa', () => {
  const CASOS: Array<[string, string, number, string]> = [
    ['es-CO', 'COP', 2_500_000, '$ 2.500.000'],
    ['es-MX', 'MXN', 2_500, '$2,500'],
    ['es-AR', 'ARS', 2_500, '$ 2.500'],
    ['es-CL', 'CLP', 2_500, '$2.500'],
    ['es-PE', 'PEN', 2_500, 'S/ 2,500'],
    ['en-US', 'USD', 2_500, '$2,500'],
    ['es-ES', 'EUR', 2_500, '2.500 €'],
  ];

  it('escribe cada moneda igual que Flutter', async () => {
    const { setMoneyLocale } = await import('./format');
    try {
      for (const [locale, currency, monto, esperado] of CASOS) {
        setMoneyLocale(locale, currency);
        expect(formatMoney(monto)).toBe(esperado);
      }
    } finally {
      setMoneyLocale('es-CO', 'COP');
    }
  });
});

File: mobile/test/domain/money/format_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:my_finance/domain/money/format.dart';

void main() {
  setUp(() => setMoneyLocale('es-CO', 'COP'));

  group('formatMoney', () {
    test('usa formato colombiano con punto de miles', () {
      expect(formatMoney(2500000), r'$ 2.500.000');
    });

    test('no muestra decimales', () {
      expect(formatMoney(9900), r'$ 9.900');
      expect(formatMoney(175094), r'$ 175.094');
    });

    test('maneja cero y negativos', () {
      expect(formatMoney(0), r'$ 0');
      expect(formatMoney(-48000), r'-$ 48.000');
    });
  });

  group('parseMoney', () {
    test('acepta lo que el usuario realmente escribe', () {
      expect(parseMoney('85000'), 85000);
      expect(parseMoney('85.000'), 85000);
      expect(parseMoney(r'$ 85.000'), 85000);
      expect(parseMoney('1.500.000'), 1500000);
    });

    test('devuelve null cuando no hay número', () {
      expect(parseMoney(''), isNull);
      expect(parseMoney('abc'), isNull);
    });
  });

  group('formatCompact', () {
    test('abrevia millones y miles', () {
      expect(formatCompact(2500000), r'$ 2,5 M');
      expect(formatCompact(85000), r'$ 85 k');
    });
  });

  group('setMoneyLocale', () {
    test('cambia la moneda de todos los formatos', () {
      setMoneyLocale('es-ES', 'EUR');
      expect(formatMoney(2500), '2.500 €');
      expect(currencySymbol(), '€');
      expect(formatCompact(2500000), '€ 2,5 M');
    });

    test('acepta el locale con guion, como lo guarda la app web', () {
      setMoneyLocale('en-US', 'USD');
      expect(formatMoney(2500), r'$2,500');
    });

    // Los mismos ejemplos que muestra la configuración inicial de la app
    // web (src/domain/money/currencies.ts). Si cambian allá y no acá, las
    // dos apps le muestran al usuario la misma plata escrita distinto.
    test('coincide con los ejemplos de la app web', () {
      const esperados = {
        ('es-CO', 'COP', 2500000): r'$ 2.500.000',
        ('es-MX', 'MXN', 2500): r'$2,500',
        ('es-AR', 'ARS', 2500): r'$ 2.500',
        ('es-CL', 'CLP', 2500): r'$2.500',
        ('es-PE', 'PEN', 2500): 'S/ 2,500',
        ('en-US', 'USD', 2500): r'$2,500',
        ('es-ES', 'EUR', 2500): '2.500 €',
      };
      esperados.forEach((clave, esperado) {
        final (loc, cur, monto) = clave;
        setMoneyLocale(loc, cur);
        expect(formatMoney(monto), esperado, reason: '\$loc/\$cur');
      });
    });

    test('vuelve a colombiano al restaurar', () {
      setMoneyLocale('es-CO', 'COP');
      expect(formatMoney(2500000), r'$ 2.500.000');
    });
  });
}
